src/par_iter/internal.rs

```diff
@@ -9,133 +9,126 @@ use super::len::*;
 
     /// Split into two producers; one produces items `0..index`, the
     /// other `index..N`. Index must be less than `N`.
-    unsafe fn split_at(self, index: usize) -> (Self, Self);
-
-    /// Unless a panic occurs, expects to be called *exactly N times*.
-    unsafe fn produce(&mut self, shared: &Self::Shared) -> Self::Item;
+    fn split_at(self, index: usize) -> (Self, Self);
 }
 
-/// A consumer which consumes items within the lifetime `'consume`.
-pub trait Consumer<'consume>: Send {
-    type Item: 'consume;
-    type Shared: Sync + 'consume;
-    type SeqState;
+/// A consumer which consumes items that are fed to it.
+pub trait Consumer<Item>: Send {
+    type Folder: Folder<Item, Result=Self::Result>;
+    type Reducer: Reducer<Self::Result>;
     type Result: Send;
 
     /// If it costs `producer_cost` to produce the items we will
     /// consume, returns cost adjusted to account for consuming them.
-    fn cost(&mut self, shared: &Self::Shared, producer_cost: f64) -> f64;
+    fn cost(&mut self, producer_cost: f64) -> f64;
 
     /// Divide the consumer into two consumers, one processing items
-    /// `0..index` and one processing items from `index..`.
-    unsafe fn split_at(self, shared: &Self::Shared, index: usize) -> (Self, Self);
+    /// `0..index` and one processing items from `index..`. Also
+    /// produces a reducer that can be used to reduce the results at
+    /// the end.
+    fn split_at(self, index: usize) -> (Self, Self, Self::Reducer);
+
+    /// Convert the consumer into a folder that can consume items
+    /// sequentially, eventually producing a final result.
+    fn fold(self) -> Self::Folder;
 
-    /// Start processing items. This can return some sequential state
-    /// that will be threaded through as items are consumed.
-    unsafe fn start(&mut self, shared: &Self::Shared) -> Self::SeqState;
+}
+
+pub trait Folder<Item> {
+    type Result;
 
     /// Consume next item and return new sequential state.
-    unsafe fn consume(&mut self,
-                      shared: &Self::Shared,
-                      state: Self::SeqState,
-                      item: Self::Item)
-                      -> Self::SeqState;
+    fn consume(self, item: Item) -> Self;
 
     /// Finish consuming items, produce final result.
-    unsafe fn complete(self, shared: &Self::Shared, state: Self::SeqState) -> Self::Result;
+    fn complete(self) -> Self::Result;
+}
+
+pub trait Reducer<Result> {
 
     /// Reduce two final results into one; this is executed after a
     /// split.
-    unsafe fn reduce(shared: &Self::Shared,
-                     left: Self::Result,
-                     right: Self::Result)
-                     -> Self::Result;
+    fn reduce(self, left: Result, right: Result) -> Result;
 }
 
 /// A stateless consumer can be freely copied.
-pub trait UnindexedConsumer<'c>: Consumer<'c> {
+pub trait UnindexedConsumer<ITEM>: Consumer<ITEM> {
     fn split(&self) -> Self;
```

Comment:
Since this does not have the type signature `fn split(self) -> (Self, Self)`, perhaps `split_off` would be better?